### PR TITLE
feat(mpnn): report per-design confidence in inference outputs

### PR DESCRIPTION
**Pull request by daylight-00**
## Summary
MPNN inference currently reports only sequence recovery, which needs a native sequence and is meaningless for de novo design. The original ProteinMPNN/LigandMPNN additionally report a per-sequence confidence (exp(-mean NLL)), the standard metric for ranking/filtering designs. This PR brings that to the foundry re-implementation.

## What's added
For each design, inference now reports:

- **confidence**: exp(-mean_over_designed_residues(log_probs)), in (0, 1], higher = more confident.
- **ligand_interface_confidence**: same, restricted to polymer-ligand interface residues (ligand_mpnn only); omitted when there are no interface residues.
- **per-residue confidence**: exp(-NLL) per position (0 at non-designed positions), stored in a dedicated `mpnn_confidence` atom-array field.

### Output format

- FASTA header: `>name_b{b}_d{d}, confidence=..., ligand_interface_confidence=..., sequence_recovery=..., ligand_interface_sequence_recovery=...`
- CIF:
  - `_mpnn_output.confidence` / `_mpnn_output.ligand_interface_confidence` (per-design scalars).
  - per-residue confidence in a dedicated `_atom_site.mpnn_confidence` column (via the existing `extra_fields` path, like `mpnn_temperature`); the input `b_factor` annotation is preserved.

## Implementation

- `metrics/nll.py`:
  - `SampledNLL` / `SampledLigandInterfaceNLL`: score the *sampled* sequence (`S_sampled`) under `log_softmax(logits)`, reusing the existing NLL math and interface-mask machinery. Note: `decoder_features["log_probs"]` is temperature/bias-scaled (default `T=0.1`), which would pin confidence near 1.0, so the raw `logits` are used to match LigandMPNN's `T=1.0` definition.
  - `MPNNConfidence` / `MPNNLigandInterfaceConfidence`: thin wrappers that expose the derived confidence (`exp(-NLL)`), so the NLL-to-confidence transform lives in the metric layer rather than the engine. Registered as `confidence` / `ligand_interface_confidence`.
- `user_settings.py`: expose `logits` in the minimal-return decoder features.
- `inference_engines/mpnn.py`: register the confidence metrics, read the per-design / per-residue confidence directly, and store per-residue confidence in a dedicated `mpnn_confidence` atom-array annotation (leaving `b_factor` untouched). When there are no polymer-ligand interface residues the interface confidence is omitted (None) rather than written as NaN.
- `utils/inference.py`: emit the confidence header fields in `write_fasta` and add `mpnn_confidence` to the CIF `extra_fields`.

## Tests

- `tests/test_metrics.py`: self-contained unit tests for the new metrics (no structure fixtures, so they run without the test-data assets the integration suite needs). They pin: the metrics read the sampled sequence and the raw logits (not the native sequence / temperature-scaled log_probs); the confidence equals `exp(-NLL)` of the sampled sequence under `log_softmax(logits)`; the interface confidence is restricted to interface residues; and the no-interface case is cleanly undefined (not a crash).

## Verification
Ran `ligand_mpnn` (legacy weights `ligandmpnn_v_32_010_25.pt`) on a dsDNA binder backbone:

```text
>na_binder_design_dsDNA_basic_0_model_0.cif_b0_d0, confidence=0.4741, ligand_interface_confidence=0.4091, sequence_recovery=0.5082, ligand_interface_sequence_recovery=0.2895
```

In the output CIF, the input `_atom_site.B_iso_or_equiv` is preserved and per-residue confidence is written to a separate `_atom_site.mpnn_confidence` column (range ~0.14-0.93 on designed positions, 0.0 on ligand/DNA/fixed atoms).

## Notes / scope

- Per-residue confidence is stored in a dedicated `mpnn_confidence` field rather than `b_factor`, so the input structure's B-factors are preserved. (The original LigandMPNN overwrites PDB b-factors; Foundry keeps them.)
- `ligand_interface_confidence` is omitted when the input has no polymer-ligand interface residues (e.g. ligand_mpnn on a monomer).
- Confidence values are stochastic across runs unless a `seed` is set.
- FASTA/CIF output overlaps with the `fix/mpnn-fasta-output` branch; confidence values live in `output_dict`, so they thread into that branch's writer by adding the two fields to its header allowlist.

### CODE REVIEW

File: models/mpnn/src/mpnn/metrics/nll.py [hunk 1]
```diff
@@ -367,3 +367,83 @@ def compute(self, log_probs, S, mask_for_loss, atom_array, **kwargs):
...
+    logits). This matches the "overall_confidence" reported by the original
+    LigandMPNN, where ``confidence = exp(-mean_nll)``.
+    """
+
+    @property
+    def kwargs_to_compute_args(self):
+        mapping = super().kwargs_to_compute_args
+        mapping["S"] = ("network_output", "decoder_features", "S_sampled")
+        mapping["log_probs"] = ("network_output", "decoder_features", "logits")
+        return mapping
```
[the remark this answers is not in the document]
Comment (daylight-00) [reply models/mpnn/src/mpnn/metrics/nll.py]: `NLL`/`InterfaceNLL`'s `kwargs_to_compute_args` is a plain `@property` that returns a fresh dict literal on every access (not cached or shared), so mutating the returned mapping cannot leak into the parent classes. No change needed.

File: models/mpnn/src/mpnn/metrics/nll.py [hunk 2]
```diff
@@ -367,3 +367,83 @@ def compute(self, log_probs, S, mask_for_loss, atom_array, **kwargs):
...
+    residues only), computed on the un-temperatured logits of the sampled
+    sequence.
+    """
+
+    @property
+    def kwargs_to_compute_args(self):
+        mapping = super().kwargs_to_compute_args
+        mapping["S"] = ("network_output", "decoder_features", "S_sampled")
+        mapping["log_probs"] = ("network_output", "decoder_features", "logits")
+        return mapping
```
[the remark this answers is not in the document]
Comment (daylight-00) [reply models/mpnn/src/mpnn/metrics/nll.py]: `NLL`/`InterfaceNLL`'s `kwargs_to_compute_args` is a plain `@property` that returns a fresh dict literal on every access (not cached or shared), so mutating the returned mapping cannot leak into the parent classes. No change needed.

File: models/mpnn/src/mpnn/metrics/nll.py [hunk 3]
```diff
@@ -367,3 +367,83 @@ def compute(self, log_probs, S, mask_for_loss, atom_array, **kwargs):
...
+
+    def compute(self, log_probs, S, mask_for_loss, **kwargs):
```
[the remark this answers is not in the document]
Comment (daylight-00) [reply models/mpnn/src/mpnn/metrics/nll.py]: Done. Renamed the `compute` parameter and its `kwargs_to_compute_args` key from `log_probs` to `logits` in both `SampledNLL` and `SampledInterfaceNLL`, so the argument name matches the raw logits it carries.

File: models/mpnn/src/mpnn/metrics/nll.py [hunk 4]
```diff
@@ -367,3 +367,83 @@ def compute(self, log_probs, S, mask_for_loss, atom_array, **kwargs):
...
+
+        Returns:
+            dict: The NLL / confidence metrics from ``NLL.compute``.
+        """
+        # Raw logits -> true log-probs at temperature 1.0 (no bias), matching
+        # LigandMPNN's confidence definition.
+        log_probs = torch.log_softmax(log_probs, dim=-1)
+        return super().compute(
+            log_probs=log_probs, S=S, mask_for_loss=mask_for_loss, **kwargs
+        )
```
[the remark this answers is not in the document]
Comment (daylight-00) [reply models/mpnn/src/mpnn/metrics/nll.py:408]: Done. Renamed the `compute` parameter and its `kwargs_to_compute_args` key from `log_probs` to `logits` in both `SampledNLL` and `SampledInterfaceNLL`, so the argument name matches the raw logits it carries.

File: models/mpnn/src/mpnn/metrics/nll.py [hunk 5]
```diff
@@ -367,3 +367,83 @@ def compute(self, log_probs, S, mask_for_loss, atom_array, **kwargs):
...
+    """
+
+    @property
+    def kwargs_to_compute_args(self):
+        mapping = super().kwargs_to_compute_args
+        mapping["S"] = ("network_output", "decoder_features", "S_sampled")
+        mapping["log_probs"] = ("network_output", "decoder_features", "logits")
+        return mapping
+
+    def compute(self, log_probs, S, mask_for_loss, atom_array, **kwargs):
```
[the remark this answers is not in the document]
Comment (daylight-00) [reply models/mpnn/src/mpnn/metrics/nll.py]: Done. Renamed the `compute` parameter and its `kwargs_to_compute_args` key from `log_probs` to `logits` in both `SampledNLL` and `SampledInterfaceNLL`, so the argument name matches the raw logits it carries.

File: models/mpnn/src/mpnn/metrics/nll.py [hunk 6]
```diff
@@ -367,3 +367,83 @@ def compute(self, log_probs, S, mask_for_loss, atom_array, **kwargs):
...
+        # Raw logits -> true log-probs at temperature 1.0 (no bias), matching
+        # LigandMPNN's confidence definition.
+        log_probs = torch.log_softmax(log_probs, dim=-1)
+        return super().compute(
+            log_probs=log_probs,
+            S=S,
+            mask_for_loss=mask_for_loss,
+            atom_array=atom_array,
+            **kwargs,
+        )
```
[the remark this answers is not in the document]
Comment (daylight-00) [reply models/mpnn/src/mpnn/metrics/nll.py:454]: Done. Renamed the `compute` parameter and its `kwargs_to_compute_args` key from `log_probs` to `logits` in both `SampledNLL` and `SampledInterfaceNLL`, so the argument name matches the raw logits it carries.

File: models/mpnn/src/mpnn/inference_engines/mpnn.py [hunk 7]
```diff
@@ -436,6 +470,19 @@ def _run_batch(
...
+            # written to the standard CIF '_atom_site.B_iso_or_equiv' column and
+            # picked up by viewers (analogous to LigandMPNN's per-residue
+            # confidence b-factors). Non-designed positions are set to 0.
+            design_confidence_atom = spread_token_wise(
+                design_non_atomized_array,
+                confidence_per_residue[design_idx],
+            )
+            full_confidence = np.zeros(len(design_atom_array), dtype=np.float32)
+            full_confidence[~design_atom_array.atomize] = design_confidence_atom
+            design_atom_array.set_annotation("b_factor", full_confidence)
```
[the remark this answers is not in the document]
Comment (daylight-00) [reply models/mpnn/src/mpnn/inference_engines/mpnn.py]: Intentional: this mirrors the original LigandMPNN, which writes per-residue confidence into the structure b-factor (non-designed positions set to 0), so viewers can color by it directly. It is called out in the PR description. We can revisit and store it under a separate annotation behind a flag if preserving input b-factors becomes important.

File: models/mpnn/src/mpnn/metrics/nll.py [hunk 8]
```diff
@@ -367,3 +367,83 @@ def compute(self, log_probs, S, mask_for_loss, atom_array, **kwargs):
...
+
+
+class SampledInterfaceNLL(InterfaceNLL):
+    """Interface NLL / confidence of the *sampled* (designed) sequence.
+
+    Mirrors LigandMPNN's "ligand_confidence" (polymer-ligand interface
+    residues only), computed on the un-temperatured logits of the sampled
+    sequence.
+    """
```
[the remark this answers is not in the document]
Comment (daylight-00) [reply models/mpnn/src/mpnn/metrics/nll.py]: Done. Added `test_sampled_interface_nll_restricts_to_interface_and_uses_sampled`, which injects a known interface mask and asserts the interface NLL only includes interface positions and scores the sampled sequence (not native `S`) under `log_softmax(logits)`. The interface-mask derivation itself is inherited from `InterfaceNLL` and covered by the integration test.

File: models/mpnn/src/mpnn/metrics/nll.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (daylight-00) [reply models/mpnn/src/mpnn/metrics/nll.py]: Done. Added `test_sampled_interface_nll_restricts_to_interface_and_uses_sampled`, which injects a known interface mask and asserts the interface NLL only includes interface positions and scores the sampled sequence (not native `S`) under `log_softmax(logits)`. The interface-mask derivation itself is inherited from `InterfaceNLL` and covered by the integration test.

File: models/mpnn/src/mpnn/metrics/nll.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (daylight-00) [reply models/mpnn/src/mpnn/metrics/nll.py:454]: Done. Added `test_sampled_interface_nll_restricts_to_interface_and_uses_sampled`, which injects a known interface mask and asserts the interface NLL only includes interface positions and scores the sampled sequence (not native `S`) under `log_softmax(logits)`. The interface-mask derivation itself is inherited from `InterfaceNLL` and covered by the integration test.

**Comment by daylight-00** [comment]
Resolves #239

**Review by AndrewKubaney — CHANGES_REQUESTED** [review]
Thanks for implementing this! It will be useful to have the ProteinMPNN/LigandMPNN confidences included in the output.

I have two changes/questions:

1. I would recommend adding `MPNNConfidence` and `MPNNInterfaceConfidence` metrics, corresponding to `overall_confidence` and `ligand_confidence` here. I think those names would make the metrics clearer and help distinguish them from other confidence metrics. These could be wrappers around the sampled NLL metrics you added, which would also move the NLL-to-confidence calculation out of the top-level `MPNNInferenceEngine`.

2. I agree that it would be useful to store the per-residue confidences in the atom array. I have some changes in another branch that could enable writing these per-residue confidences to the output CIF. My only concern is that overwriting the b-factor may be undesirable if users want to retain the original b-factor annotation. Could we instead create a separate atom array field for the MPNN confidence? If the goal is to write this into the CIF output, we can also discuss how to merge this with the changes in my `fix/mpnn-fasta-output` branch.

Happy to discuss further!

-Andrew

**Comment by daylight-00** [comment]
Thanks for the review, Andrew. I've addressed both points.

1. I added MPNNConfidence / MPNNInterfaceConfidence wrappers so the sampled NLL metrics and the derived exp(-NLL) confidence values stay separated. The inference engine now consumes the confidence outputs directly instead of performing the NLL-to-confidence transform itself.

2. I also changed the per-residue confidence storage to avoid overwriting b_factor. I originally used b_factor because the original LigandMPNN writes per-residue confidence there, but I agree that preserving the input B-factor is cleaner for Foundry. The values now go into a dedicated `mpnn_confidence` AtomArray annotation.

For CIF output, I added `mpnn_confidence` to the existing `extra_fields` allowlist, so it serializes as `_atom_site.mpnn_confidence`, following the same path as other MPNN-specific fields like `mpnn_temperature`. I took a look at your `fix/mpnn-fasta-output` branch, and this seems compatible with the `write_structure` changes there, but I'm happy to rename/re-categorize the field if that branch settles on a different output convention.

For the FASTA output overlap, I also checked the refactored writer in your branch. Since the confidence values are now in `output_dict`, I think they should thread into the new design-entry writer by adding `overall_confidence` and `ligand_confidence` to the design-entry header allowlist. Let me know whether you'd prefer me to rebase onto your branch or land this first and reconcile afterward.

David

**Review by AndrewKubaney — CHANGES_REQUESTED** [review]
Apologies for the delay in my review! This looks great; thank you for implementing this! I would suggest a few minor changes:

1) For the `kwargs_to_compute_args` function in your new metrics classes, I think it would be more straightforward to construct the mapping dictionary from scratch, rather than relying on `super`. This will avoid the `.pop` logic, which might be hard to understand.

2) I think `MPNNInterfaceConfidence` could be renamed to `MPNNLigandInterfaceConfidence`, so that people don't misinterpret this as applying to non protein-ligand interfaces (for instance, protein-protein interfaces).

3) I think `mpnn_confidence` is a good name for the atom array feature; what do you think about changing the final confidence names (the ones that ends up in the FASTA) from `overall_confidence` to `confidence` or `mpnn_confidence` (and similarly `ligand_confidence` to `ligand_interface_confidence` or `mpnn_ligand_interface_confidence`)? I think those might be more descriptive, but they do deviate from the original ProteinMPNN/LigandMPNN norm.

4) One edge case involves attempting to compute the ligand interface confidence when no ligand is present (if someone runs LigandMPNN on a protein monomer) or there are no valid ligand-interface residues. Could you add a test case to make sure this is handled cleanly?

Thanks again for your work on this! After addressing these, I think we should merge. The other branch can be merged at a later time.

- Andrew

**Comment by daylight-00** [comment]
Thanks, Andrew — all four addressed.

1. `kwargs_to_compute_args`: switched to constructing the mapping explicitly. (I'd used a copy + `pop("log_probs")` to avoid mutating the parent mapping, but I agree the explicit literal is clearer.)

2. Renamed `MPNNInterfaceConfidence` -> `MPNNLigandInterfaceConfidence` (and the underlying `SampledInterfaceNLL` -> `SampledLigandInterfaceNLL`) so the polymer-ligand scope is explicit.

3. Renamed the reported fields: `overall_confidence` -> `confidence` and `ligand_confidence` -> `ligand_interface_confidence`. I like this better too — it mirrors the existing `sequence_recovery` / `ligand_interface_sequence_recovery` keys and Foundry's un-prefixed output-key convention (the `mpnn_` prefix is used for atom-array fields like `mpnn_confidence`, which I kept as-is).

4. Added handling + a test for the no-interface case: when `ligand_mpnn` runs on an input with no polymer-ligand interface residues, the interface confidence is undefined (`NaN`, flagged by `valid_examples_mask`), and the engine now converts it to `None` so it's omitted rather than written as `NaN`.

One small naming question: the sibling metric classes in mpnn/metrics are unprefixed (`NLL`, `SequenceRecovery`, `InterfaceNLL`, ...), so for consistency I could drop the `MPNN` prefix on the new classes -> `Confidence` / `LigandInterfaceConfidence`. I'm equally happy to keep `MPNN*` if you prefer it.

Sounds good on merging this and leaving the fix/mpnn-fasta-output integration for later. Thanks for the thorough review!

David

**Review by AndrewKubaney — APPROVED** [review]
Sorry for the delay, and thanks for your patience! This looks good to merge. The `MPNN` prefix on the classes is okay to keep for now.

I ran the MPNN tests locally, and your changes do not break any of the tests. There are a few stale unrelated tests, but your new tests pass.

Thanks again for your contribution!

Andrew
